core/ipc/rpc/kafka/src/main/java/org/opennms/core/ipc/rpc/kafka/KafkaRpcClientFactory.java

```diff
@@ -0,0 +1,311 @@
+            while (true) {
+                try {
+                    ResponseCallback responseCb = delayQueue.take();
+                    responseCb.sendResponse(null);
+                } catch (InterruptedException e) {
+                    // exit, nothing to do
+                }
+            }
+        });
+    }
+
+    /** Handle Response from kafka consumer and timeout thread **/
+    private class ResponseHandler<S extends RpcRequest, T extends RpcResponse> implements ResponseCallback {
+
+        private final CompletableFuture<T> responseFuture;
+        private final RpcModule<S, T> rpcModule;
+        private final long expirationTime;
+        private final String rpcId;
+ 
+        public ResponseHandler(CompletableFuture<T> responseFuture, RpcModule<S, T> rpcModule, String rpcId,
+                long timeout) {
+            this.responseFuture = responseFuture;
+            this.rpcModule = rpcModule;
+            this.expirationTime = timeout;
+            this.rpcId = rpcId;
+        }
+
+        @Override
+        public void sendResponse(String message) {
+            T response = null;
+            // When message is not null, it's called from kafka consumer otherwise it is from timeout tracker.
+            if (message != null) {
+                response = rpcModule.unmarshalResponse(message);
+                if (response.getErrorMessage() != null) {
+                    responseFuture.completeExceptionally(new RemoteExecutionException(response.getErrorMessage()));
+                } else {
+                    responseFuture.complete(response);
+                }
+            } else {
+                responseFuture.completeExceptionally(new RequestTimedOutException(new TimeoutException()));
+            }
+            rpcResponseMap.remove(rpcId);
+        }
+
+        @Override
+        public int compareTo(Delayed other) {
+            long myDelay = getDelay(TimeUnit.MILLISECONDS);
+            long otherDelay = other.getDelay(TimeUnit.MILLISECONDS);
+            if (myDelay < otherDelay)
+                return -1;
+            if (myDelay == otherDelay)
+                return 0;
+            return 1;
+        }
+
+        @Override
+        public long getDelay(TimeUnit unit) {
+            long now = System.currentTimeMillis();
+            return unit.convert(expirationTime - now, TimeUnit.MILLISECONDS);
+        }
+
+    }
+
+    private class KafkaConsumerRunner implements Runnable {
+        private final KafkaConsumer<String, byte[]> consumer;
+        private final AtomicBoolean closed = new AtomicBoolean(false);
+
+        public KafkaConsumerRunner(KafkaConsumer<String, byte[]> consumer) {
+            this.consumer = consumer;
+        }
+
+        @Override
+        public void run() {
+            try {
+                // subscribe to topics with "rpc-response" as part of topic name.
+                Pattern pattern = Pattern.compile(".*rpc-response.*");
```

Comment:
Won't this match topics that have 'rpc-response' *anywhere* in the name?

Reply:
Yeah, if there are multiple instances of OpenNMS and only kafka instance, this may be a problem.
 I will add instance name to regex match